### PR TITLE
Make static final buffers unreleasable and read-only

### LINKED ISSUE #11792: Data corruption due to interaction between `SslHandler` and `HttpObjectEncoder`
TL;DR:

`HttpObjectEncoder` sends a duplicate of its static final `ByteBuf` field to encode HTTP chunks, such as `0\r\n\r\n`. This reused buffer is neither read-only nor unexpandable, and thus anyone in the pipeline can modify or expand it. As a result, `SslHandler` can sometimes expand it and append stuff to it, causing unexpected data corruption. It'd be nice if this issue is fixed soon and a new release is out given its severity.

Here's the full analysis from my colleague @alexc-db:

1. When `Channel.flush()` is called, `SslHandler` starts getting items from the `pendingUnencryptedWrites` queue [1] - it gets entries from the internal `bufAndListenerPairs` and accumulates them in the `toReturn` buffer [2].
2. For the first entry, `toReturn` is `null` and thus `composeFirst` is called. That method by default just returns the reference to the first entry (it’s overridden in `SslHandlerCoalescingBufferQueue` to have special handling for `CompositeByteBuf`, but it’s not relevant to the issue here). The following calls are calling `compose` and pass `toReturn` there.
3. Internally, `compose` calls `attemptCopyToCumulation` which checks if `toReturn` is writable and its capacity can be expanded to write new bytes from the following queued buffers - if it passes checks in [3], `attemptCopyToCumulation` copies bytes from the next entries to the original `toReturn` buffer (as we saw above it can be the first entry), if it does not pass the checks - `SslHandlerCoalescingBufferQueue` calls `copyAndCompose` which creates a new bigger `buf` and copies bytes from `toReturn` and next entry there, making it a new `toReturn`.
4. Given the behavior above - if `flush` is called and `bufAndListenerPairs` has multiple buffers queued where the first entry is writable, had enough capacity for the bytes in the next entries, and shared across multiple threads - two threads will start copying bytes into the same memory addresses which will cause corruption. In our local reproducer that has more dependencies that just Netty, it was `ZERO_CRLF_CRLF_BUF` [4] (`duplicate()` call doesn’t make a copy, just wraps the original direct buffer).
I was able to make tests pass by marking `ZERO_CRLF_CRLF_BUF read-only` [5].
[1] https://github.com/netty/netty/blob/netty-4.1.68.Final/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L817
[2] https://github.com/netty/netty/blob/netty-4.1.68.Final/transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java#L175
[3] https://github.com/netty/netty/blob/netty-4.1.68.Final/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L2326-L2334
[4] https://github.com/netty/netty/blob/netty-4.1.68.Final/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java#L210
[5]
```diff
diff --git a/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java b/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java
--- a/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java (revision 7d34282f9d2ffdd64c91cb4780b09902d9779b92)
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java (date 1635112426650)
@@ -50,9 +50,9 @@
     static final int CRLF_SHORT = (CR << 8) | LF;
     private static final int ZERO_CRLF_MEDIUM = ('0' << 16) | CRLF_SHORT;
     private static final byte[] ZERO_CRLF_CRLF = { '0', CR, LF, CR, LF };
-    private static final ByteBuf CRLF_BUF = unreleasableBuffer(directBuffer(2).writeByte(CR).writeByte(LF));
+    private static final ByteBuf CRLF_BUF = unreleasableBuffer(directBuffer(2).writeByte(CR).writeByte(LF).asReadOnly());
     private static final ByteBuf ZERO_CRLF_CRLF_BUF = unreleasableBuffer(directBuffer(ZERO_CRLF_CRLF.length)
-            .writeBytes(ZERO_CRLF_CRLF));
+            .writeBytes(ZERO_CRLF_CRLF).asReadOnly());
     private static final float HEADERS_WEIGHT_NEW = 1 / 5f;
     private static final float HEADERS_WEIGHT_HISTORICAL = 1 - HEADERS_WEIGHT_NEW;
     private static final float TRAILERS_WEIGHT_NEW = HEADERS_WEIGHT_NEW;
```

### Netty version

4.1.68.Final and 4.1.69.Final and of course possible in older versions.

### PR DESCRIPTION
**Pull request by alexc-db**
Motivation:

Fix https://github.com/netty/netty/issues/11792 and avoid potential corruption issues caused by writable shared buffers.

TL;DR:

`HttpObjectEncoder` sends a duplicate of its static final `ByteBuf` field to encode HTTP chunks, such as `0\r\n\r\n`. This reused buffer is neither read-only nor unexpandable, and thus anyone in the pipeline can modify or expand it. As a result, `SslHandler` can sometimes expand it and append stuff to it, causing unexpected data corruption.

Modification:

Most of the static final buffers in the project are made unreleasable and read-only by wrapping them in `Unpooled.unreleasableBuffer(...).asReadOnly()`. Buffers that were excluded:

- Unpooled.EMPTY_BUFFER - EmptyByteBuf instances are not readable/writable/releasable. Wrapping this buffer in `Unpooled.unreleasableBuffer(...).asReadOnly()` would change the instance type, since it's a public API it could be a potential breaking change.
- ReplayingDecoderByteBuf.EMPTY_BUFFER - since it uses Unpooled.EMPTY_BUFFER and not releasable too.

Excluded buffers are marked with `SuppressWarnings` to skip check added in https://github.com/netty/netty-build/pull/12.

Result:

Fixes #11792.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/Unpooled.java [hunk 1]
```diff
@@ -87,6 +87,7 @@
     /**
      * A buffer whose capacity is {@code 0}.
      */
+    @SuppressWarnings("checkstyle:StaticFinalBuffer")  // EmptyByteBuf is not writeable or readable.
     public static final ByteBuf EMPTY_BUFFER = ALLOC.buffer(0, 0);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/Unpooled.java:91]: got it... ok yeah its ok then :)

**Comment by normanmaurer** [comment]
@alexc-db can you sign our ICLA: https://netty.io/s/icla ?

**Comment by alexc-db** [comment]
@normanmaurer our legal works on signing the CCLA, I'll check with them.

In the meantime - I noticed that you released netty-build with https://github.com/netty/netty-build/pull/12, would you like me to include the updated version with a check in this PR or submit a separate PR afterwards?

**Comment by normanmaurer** [comment]
@alexc-db yes please

**Comment by normanmaurer** [comment]
@alexc-db let me know once the legal stuff is sorted as I would like to pull this in and do a release :)

**Comment by alexc-db** [comment]
@normanmaurer should be done as of now, please check if you can see the signed CCLA on your side. sorry for the delay :)

**Comment by normanmaurer** [comment]
@alexc-db yep its there... thanks!
